### ISSUE TITLE
[Datahub] - Add filter options on the download menu

### ISSUE THREAD

**Issue by tomalary** [geonetwork/geonetwork-ui#766]
On each datasets page, the user can access to options to filter download resources and not only per file extension. The filters can be set up (for example per year).

The filter is a drop-down menu with suggestion and autocomplete search on each drop-down menu. 

![image](https://github.com/geonetwork/geonetwork-ui/assets/108727123/a63eab91-2704-443f-8ec9-005d3d494a4f)
_Draft of the component..._
 
For the display of the download links the component check the tag in the metadata. Then we create the download links and they are displayed. If the tag are not found the component work as the usual downloading component.  

As our solution is to create the download links by using **API Entrepot** (API developed by IGN that allow to load/process/share data services), we are open to discuss about an enhancement that suits everyone. 

_If this feature is not clear please let us know or ask questions to clarify._

---

**This work is sponsored by IGN**.

**Comment by jahow** [comment]
Hi, thanks for your interest,

Could you please explain what are the tags you mention that let you apply these filters? are these standard ISO fields? Also, how many download links do you typically expect in order for such filtering to become valuable?

**Comment by IGNF-Xavier** [comment]
Hi,
_API entrepôt_ will allow the owner of datasets to configure prepackage zip-files of data (PPK) depending on territories, date of products, format of files, etc. 
Once done, these PPK can be downloaded at a specific URL thanks to _API Telechargement_
The user of the component in the dasasets page will be able to choose the PPK needed according to user-defined filters :
- polygon : x1,y1;x2,y2;… ( EPSG:4326), 
- editionDateFrom : YYYY-MM-DD, 
- editionDateTo : YYYY-MM-DD, 
- lang : FR ou EN
- crs :  epsg,
- zone : FRA (for France entière) or D075 (for Paris)...
- format : SHP (for Shapefile)...

The swagger of API telechargement is here : https://data.geopf.fr/telechargement/swagger-ui/index.html (today 16/01, it won't work because of maintenance).
The point of access of the capabilities of this download-service will be stored in the metadata of the dataset. If completed, the component appears, if not, the component is not displayed.

When you consider that there are 101 departments in France, maybe the results of this component will need to have pagination.

**Comment by jahow** [comment]
Thanks for the clarification, it definitely makes more sense! Do we agree that records with this feature would be service metadata, and not datasets?

There has been a discussion for a while about how/if we should show services in the datahub; currently they're filtered out. A component such as the one mentioned here would make sense for other kinds of services (WMS, OGC API etc.), so we could address this use case globally and do a specific implementation for the IGN APIs.

Also, could you please clarify if this work will be funded by IGN?

**Comment by IGNF-Xavier** [comment]
We (IGN) are currently developing this download component specifically for the datasets pages because there will be plenty of datasets with different years. 
For instance, a new edition of BD TOPO is available every 6 months since many years.

Maybe we will dadapt this component for this specific PPK-download service page adding the filter dataset. 

Then we (IGN) will develop a map-based download component principally used for datasets which have a high weight (for exemple LidarHD)

**Comment by jahow** [comment]
Ok, I see, then this interface would simply show up next to an API link pointing at the IGN Géoplateforme, regardless of the metadata type we're in. I think such a "capabilities explorer" component would be very useful even outside of the IGN use cases, e.g. listing layers in a WMS endpoint.

Would you agree to participate in a preliminary UX work to properly design this component and its integration in the datahub?

**Comment by jahow** [comment]
@IGNF-Xavier Could you please get in touch with someone at Camptocamp? (your contact information is not public) Thanks!

**Comment by IGNF-Xavier** [comment]
👍  @fgravin knows @tomalary and me

**Comment by jahow** [comment]
Meetings note are here (FR): https://semestriel.framapad.org/p/api-telechargement-datahub-a5o8?lang=fr

**Comment by mmohadIGN** [comment]
This is the workflow we plan to do. Did it seem alright to you? @fgravin and @jahow 

```mermaid
---
title: Workflow IGN download API
---
%% doc mermaid ici https://mermaid-js.github.io/mermaid/#/flowchart?id=flowcharts-basic-syntax
flowchart TD
    A("record-apis-component") -->|choose proxy compenent API| B("feature/record/proxy-apis-component")
    B -->|do| C(feature/record/ign-api-dl)
    C -->|begin| D(IGN API client)
    C -->|display| E(ui/ign-dl-form)
    D <-->|look at| C
```

**Comment by jahow** [comment]
Thanks for working on this and giving more details:
* I think there is no need to introduce a "proxy" component; currently the `gn-ui-record-api-form` is used exclusively but I think it would make sense to only use this component if the protocol is `ogcFeatures`, and introduce a new component called e.g. `gn-ui-record-ign-api-form` based on the IGN-specific protocol code; the choice between both can be done in the `record-apis-component`
* a dedicated `ign-dl-form` presentation component might not be needed; IIRC the component will mostly be comprised of several dropdown/search fields and a list of results; the smart component (e.g. `gn-ui-record-ign-api-form`) could hold the logic, interact with the IGN API client and place the relevant presentation components; this would avoid introducing another layer of complexity

Pinging @fgravin because there may be things that were discussed before that I don't know about :slightly_smiling_face:

**Comment by fgravin** [comment]
> * I think there is no need to introduce a "proxy" component; currently the `gn-ui-record-api-form` is used exclusively but I think it would make sense to only use this component if the protocol is `ogcFeatures`, and introduce a new component called e.g. `gn-ui-record-ign-api-form` based on the IGN-specific protocol code; the choice between both can be done in the `record-apis-component`

Where would you do the switch statement (API type => corresponding form) ? in `datahub-record-apis` ? Was just wondering if it's not more a `feature` than a `datahub` thiing.

> * a dedicated `ign-dl-form` presentation component might not be needed; IIRC the component will mostly be comprised of several dropdown/search fields and a list of results; the smart component (e.g. `gn-ui-record-ign-api-form`) could hold the logic, interact with the IGN API client and place the relevant presentation components; this would avoid introducing another layer of complexity

Yes I told them to start with implementing the feature component, then perhaps we could split it with some UI components or not. Might not be needed though.

**Comment by jahow** [comment]
> Where would you do the switch statement (API type => corresponding form) ? in `datahub-record-apis` ? Was just wondering if it's not more a `feature` than a `datahub` thiing.

Right, good point! then a generic `gn-ui-record-ign-api-form` in the feature/record lib would make sense, and the switch would be done here.

**Comment by mmohadIGN** [comment]
> Hi @mmohadIGN @tomalary, this PR might interfere with your work there? #875
> 
> We're going to merge this soon so as to make it part of the next release, let us know if there's anything that can be done to help with potential conflicts.

Hi, 

Thanks, it should be okay for us, but we will tell you if not

**Comment by mmohadIGN** [comment]
Hello @jahow ,

We’ve completed the development of the first version. This version includes a form to download data from our "API téléchargement".

Before proceeding, I would like to confirm if I should initiate a PR into the main branch. I will also link the relevant issue.

Please let me know if you need further details or if any additional steps are required before the merge.

**Comment by jahow** [comment]
Hi @mmohadIGN,

> Before proceeding, I would like to confirm if I should initiate a PR into the main branch. I will also link the relevant issue.

yes please!

> Please let me know if you need further details or if any additional steps are required before the merge.

Since this sounds like a significant issue we will go through a review process. Please make sure that the PR is correctly applied to the latest main branch, thanks!

**Comment by jahow** [comment]
Done!
